Implement highlighting and tier overlay for emerald pouch items

ekgame opened the pull request:
Implements emerald pouch item highlighting and tier overlay.

<table>
<tr>
	<td>Before</td>
	<td>After</td>
<tr>
	<td><img src="https://user-images.githubusercontent.com/1730245/188854731-4481f7fd-ce07-425a-a3ba-74787f6956f5.png">
	<td><img src="https://user-images.githubusercontent.com/1730245/188854756-8c4c2003-ef53-496c-8f3a-ac6829cdcc65.png">
</table>

This pull request also includes some related improvements:
- Improve roman numeral parsing
- Do not emit an "invalid item" error if emerald pouch item does not have a usage line. Valid emerald pouch items in the market do not have the line. It should be safe to assume that even a truly invalid item simply does not contain emerald.

Incompleteusern commented:
Try calling `ServerListManager#updateServers` before and seeing if that fixes the need for the offset

magicus commented:
Interesting idea. Do you think it would be possible to enhance this to provide an easy way to find servers with a suitable uptime for loot runs?

magicus commented:
> 



> Do you think it would be possible to enhance this to provide an easy way to find servers with a suitable uptime for loot runs?

I've been thinking a bit more about this. I think it would be great to have a `/servers` command, that lists all servers, with uptime, number of free slots and time to next sp. (I think we used to have something like this?)

Then you could all an argument, like `[all] | sp | lootrun`, so that if you specify `sp`, you got the behavior that you've implemented here, sorting and filtering the list wrt time to next soul point. And if you specify `lootrun`, you filter and sort servers based on how suitable they are for lootrun (not too young, but not likely to reboot soon either). And `all` is just the default behavior. (Can maybe be skipped)

I realize this is quite different from your original patch. If you don't think this is a good idea, or think it is but don't want to implement it yourself, just ignore what I've said. :-)

DonkeyBlaster commented:
I think the `/servers` command is a great idea, and the implementation idea you have is really clean too. However, I don't think I can implement this anytime soon, as I'm getting pretty busy with university applications. I'd be happy to hand this PR over to someone else if they would like to give it a shot.

magicus commented:
I might be able to spend time on it, at least later in december, unless someone else picks it up before that.

P0keDev commented:
> > 
> 
> 
> 
> > Do you think it would be possible to enhance this to provide an easy way to find servers with a suitable uptime for loot runs?
> 
> I've been thinking a bit more about this. I think it would be great to have a `/servers` command, that lists all servers, with uptime, number of free slots and time to next sp. (I think we used to have something like this?)
> 
> Then you could all an argument, like `[all] | sp | lootrun`, so that if you specify `sp`, you got the behavior that you've implemented here, sorting and filtering the list wrt time to next soul point. And if you specify `lootrun`, you filter and sort servers based on how suitable they are for lootrun (not too young, but not likely to reboot soon either). And `all` is just the default behavior. (Can maybe be skipped)
> 
> I realize this is quite different from your original patch. If you don't think this is a good idea, or think it is but don't want to implement it yourself, just ignore what I've said. :-)

Regarding sorting servers by lootrun quality: It wouldn't really work, since the main aspect of finding a good world to lootrun is for it to not have been looted by someone else recently. Beyond the world not being brand new, uptime isn't a particularly important factor of chest loot.

magicus commented:
@P0keDev Things might have changed since I studied the theory behind loot runs, but I learned that loot chests basically start out empty, and that the server add loots to them over time, until they are full. And the process to completely fill up the chests take >= 3 hours. Of course, if someone has looted the chests already you're out of luck, but starting your run in a too young world is also bit of pointless.

I've been doing a lot of loot runs lately, and I've started disliking the feature that Wynncraft skips the hub and just throws you into a world. Since I can't tell how old it is, I must start by going back to the hub. Then I might return to the same world, but I needed to know that it was not brand new.

Anyway, listing servers with uptime from a command while on a world seems like a good idea to me, regardless.

xcfrg commented:
> I've been doing a lot of loot runs lately, and I've started disliking the feature that Wynncraft skips the hub and just throws you into a world. Since I can't tell how old it is, I must start by going back to the hub. Then I might return to the same world, but I needed to know that it was not brand new.

Can't you check the Day counter in the F3 menu? Although I'm not arguing with the uptime command, it's a great idea

magicus commented:
@xcfrg Interesting idea, I have not even thought about it. That also means that if Wynntils want to check uptime for the current server, we can do so using Minecraft internals, and need not consult the Athena API.

DonkeyBlaster commented:
> I've been doing a lot of loot runs lately, and I've started disliking the feature that Wynncraft skips the hub and just throws you into a world. Since I can't tell how old it is, I must start by going back to the hub. Then I might return to the same world, but I needed to know that it was not brand new.

#382, though that method does take data from the Athena cache

magicus commented:
@DonkeyBlaster Oh, sorry I missed that. Merged now.

DonkeyBlaster commented:
> Try calling `ServerListManager#updateServers` before and seeing if that fixes the need for the offset

Tried, doesn't work.

DonkeyBlaster commented:
>I've been thinking a bit more about this. I think it would be great to have a /servers command, that lists all servers, with uptime, number of free slots and time to next sp. (I think we used to have something like this?)

I just realized we already have a server command, `/s`. Argument `list` just lists all the servers online, nothing special. Argument `info <server>` just lists the player in that server.

Wynncraft also uses `/server` for some internal thing, and `/servers` just brings you to hub. I think a rework of the `/s` command is the best option here.

Incompleteusern commented:
Try changing the first - to a + in `ServerProfile#matchTime` and see if that works

DonkeyBlaster commented:
Someone can either take over or we can just let this die; I don't want to work on it rn

ekgame committed:
Improve roman numeral parsing.

ekgame committed:
Do not emit a warning if an emerald pouch does not have a usage line.

Valid emerald pouch items do not have a usage line in the market, assume that they are empty by default.

ekgame committed:
Implemented emerald pouch highlight and tier overlay.

kristofbolyai commented on common/src/main/java/com/wynntils/utils/MathUtils.java:
```diff
@@ -4,9 +4,24 @@
 
 public final class MathUtils {
+
+    private static final Map<Integer,Integer> romanNumeralsMap = new HashMap<>();
```
TreeMap is better for this?

ekgame replied:
I don't really see why TreeMap would be more appropriate. The map only has 7 items, so memory usage and the order of entries is not an issue.

kristofbolyai replied:
Sure, fine either way.

kristofbolyai commented on common/src/main/java/com/wynntils/wynn/item/parsers/EmeraldPouchParser.java:
```diff
@@ -22,13 +22,7 @@ public static int getPouchUsage(ItemStack stack) {
-            WynntilsMod.error(
-                    "EmeraldPouchParser#getPouchUsage was called on an ItemStack that wasn't an emerald pouch");
-            return -1;
+            return 0;
```
This fixes inventory emerald count too. Thanks

kristofbolyai commented:
@ekgame 
```
Execution failed for task ':common:spotlessJava'.
> Do not use wildcard imports.  'spotlessApply' cannot resolve this issue.
```

Could you remove wildcard imports? Otherwise this looks good!

ekgame committed:
Revert wildcard import.

ekgame commented:
Looks like my IDE collapsed them automatically, should be good now.

ekgame committed:
chore: spotless formatting

kristofbolyai committed:
Merge branch 'main' into feature/emerald-pouch-tier-overlay

magicus commented:
Thank you for your contribution to Artemis @ekgame!

ekgame commented:
> Thank you for your contribution to Artemis @ekgame!

Happy to help!
